File: src/sandbox/runner.ts
// L4 — Sandbox. Runs LLM-generated JS in a child process with the Node
// permission model: filesystem restricted to the job's scratch dir (+ read-only
// node_modules), no child processes, no workers.
// v1 limitation (accepted): network is NOT restricted by --permission.

import { spawn } from 'child_process'
import fs from 'fs'
import path from 'path'
import { v4 as uuid } from 'uuid'

const SCRATCH = path.resolve('scratch')
const TIMEOUT_MS = 30_000

export interface SandboxResult {
  success: boolean
  stdout: string
  stderr: string
  files: string[]      // files present in job dir after run (absolute paths)
  jobDir: string
}

export async function runInSandbox(code: string, inputFiles: string[] = []): Promise<SandboxResult> {
  const jobDir = path.join(SCRATCH, 'jobs', uuid())
  fs.mkdirSync(jobDir, { recursive: true })

  // copy input files (e.g. downloaded attachments) into the job dir
  for (const f of inputFiles) {
    if (fs.existsSync(f)) fs.copyFileSync(f, path.join(jobDir, path.basename(f)))
  }

  const scriptPath = path.join(jobDir, 'script.mjs')
  fs.writeFileSync(scriptPath, code, 'utf8')

  const nodeModules = path.resolve('node_modules')

  const child = spawn(
    process.execPath,
    [
      '--permission',
      `--allow-fs-read=${jobDir}`,
      `--allow-fs-read=${nodeModules}`,
      `--allow-fs-write=${jobDir}`,
      scriptPath,
    ],
    { cwd: jobDir, timeout: TIMEOUT_MS, windowsHide: true }
  )

  let stdout = ''
  let stderr = ''
  child.stdout.on('data', d => (stdout += d.toString()))
  child.stderr.on('data', d => (stderr += d.toString()))

  const exitCode: number = await new Promise(resolve => {
    child.on('close', code => resolve(code ?? 1))
    child.on('error', () => resolve(1))
  })

  const files = fs
    .readdirSync(jobDir)
    .filter(f => f !== 'script.mjs')
    .map(f => path.join(jobDir, f))

  return {
    success: exitCode === 0,
    stdout: stdout.slice(0, 8000),
    stderr: stderr.slice(0, 4000),
    files,
    jobDir,
  }
}

File: python/app/sandbox/runner.py
"""L4 — Sandbox. Runs LLM-generated Python in an isolated subprocess
(-I: isolated mode, ignores user site-packages and env vars), cwd locked to a
per-job scratch dir, 30s timeout.

v1 limitation (accepted, same class as the Node port's): the OS does not
restrict filesystem or network for the child. Containment is cwd + isolated
mode + timeout. Move to Docker/E2B when code-exec risk grows."""

import shutil
import subprocess
import sys
import uuid
from dataclasses import dataclass, field
from pathlib import Path

from ..config import SCRATCH_DIR

TIMEOUT_S = 30


@dataclass
class SandboxResult:
    success: bool
    stdout: str
    stderr: str
    files: list[str] = field(default_factory=list)   # files in job dir after run
    job_dir: str = ""


def run_in_sandbox(code: str, input_files: list[str] | None = None) -> SandboxResult:
    job_dir = SCRATCH_DIR / "jobs" / uuid.uuid4().hex
    job_dir.mkdir(parents=True, exist_ok=True)

    for f in input_files or []:
        src = Path(f)
        if src.exists():
            shutil.copyfile(src, job_dir / src.name)

    script_path = job_dir / "script.py"
    script_path.write_text(code, encoding="utf-8")

    try:
        proc = subprocess.run(
            [sys.executable, "-I", str(script_path)],
            cwd=job_dir,
            capture_output=True,
            text=True,
            timeout=TIMEOUT_S,
        )
        exit_code = proc.returncode
        stdout, stderr = proc.stdout, proc.stderr
    except subprocess.TimeoutExpired as e:
        exit_code = 1
        stdout = (e.stdout or b"").decode() if isinstance(e.stdout, bytes) else (e.stdout or "")
        stderr = f"Timed out after {TIMEOUT_S}s"

    files = [str(p) for p in job_dir.iterdir() if p.name != "script.py"]
    return SandboxResult(
        success=exit_code == 0,
        stdout=stdout[:8000],
        stderr=stderr[:4000],
        files=files,
        job_dir=str(job_dir),
    )
